### PR TITLE
Overhaul LLDB Tutorial

### PR DESCRIPTION
**Pull request by awdeorio**
Overhaul the LLDB tutorial to make the example in sync with recent updates to our IDE tutorials.

### CODE REVIEW

**Comment by awdeorio** [comment]
Bump @ohjuny :)

**Comment by ohjuny** [comment]
> Bump @ohjuny :)

Just saw this! Will make sure to review this and #121 over the weekend!

File: docs/setup_lldb.md [hunk 1]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
+`s` **Step AKA Step Into**
+Run one line of code, stepping _into_ any function calls to execute them line-by-line.
+
+`up` **Up AKA Step Out**
+Run the program until it returns from the current function (or until the next breakpoint).
+
+`c` **Continue**
+Run the program until the next breakpoint.
+
+`q` **Quit**
```
Comment (ohjuny) [inline docs/setup_lldb.md:136]:
```suggestion
`q` **Quit:**
```

File: docs/setup_lldb.md [hunk 2]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
+
+To get started, copy this example `main.cpp` into your editor.
+```c++
+#include <iostream>
+#include <vector>
+using namespace std;
+
+double sum (const vector<double> &data) {
+  double total = 0;
+  for (size_t i=0; i<data.size(); ++i) {
```
Comment (ohjuny) [inline docs/setup_lldb.md:149]:
```suggestion
  for (size_t i = 0; i < data.size(); ++i) {
```
nitpicking....

File: docs/setup_lldb.md [hunk 3]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
 
-**Pro-tip**: Other ways to set breakpoints:
+<div class="primer-spec-callout info" markdown="1">
+**Pro-tip**: There are several ways to set breakpoints.
 
 | `b main` | breakpoint on main function |
-| `b 29` | breakpoint on line 29 of current file |
-| `b stats.cpp:16` | breakpoint on line 16 of stats.cpp |
+| `b 13` | breakpoint on line 13 of current file |
+| `b main.cpp:13` | breakpoint on line 16 of stats.cpp |
```
Comment (ohjuny) [inline docs/setup_lldb.md:188]:
```suggestion
| `b main.cpp:13` | breakpoint on line 13 of main.cpp |
```

File: docs/setup_lldb.md [hunk 4]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
+-> 14  	  vector<double> data;
+   15  	  data.push_back(10);
+   16  	  data.push_back(20);
+   17  	  data.push_back(30);
+Target 0: (main.exe) stopped.
+```
+{: data-highlight="1,9" }
+
+### Step over
+Enter `n` AKA "Next" AKA "Step Over" a few times until you reach the highlighted line of code.
```
Comment (ohjuny) [inline docs/setup_lldb.md:213]:
```suggestion
Enter `n` (Next / Step Over) a few times until you reach the highlighted line of code.
```
Refer to AKA comment above.

File: docs/setup_lldb.md [hunk 5]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
+   15  	  data.push_back(10);
+   16  	  data.push_back(20);
+   17  	  data.push_back(30);
+-> 18  	  cout << "sum(data) = " << sum(data) << endl;
+   19  	}
 ```
+{: data-highlight="1,8" }
+
+<div class="primer-spec-callout info" markdown="1">
+**Pro-tip:** Hit Return to repeat your previous command.
```
Comment (ohjuny) [inline docs/setup_lldb.md:229]:
```suggestion
**Pro-tip:** Hit <kbd>Return</kbd> to repeat your previous command.
```
Styling suggestion.

File: docs/setup_lldb.md [hunk 6]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
+(std::vector<double, std::allocator<double> >) $0 = size=3 {
+  [0] = 10
+  [1] = 20
+  [2] = 30
+}
 ```
+{: data-highlight="1" }
+
+### Step into
+Enter `s` AKA "Step" AKA "Step Into".  The cursor enters the `sum()` function.
```
Comment (ohjuny) [inline docs/setup_lldb.md:246]:
```suggestion
Enter `s` (Step / Step Into).  The cursor enters the `sum()` function.
```
Refer to AKA comment above.

File: docs/setup_lldb.md [hunk 7]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
+   5   	double sum (const vector<double> &data) {
+-> 6   	  double total = 0;
+   7   	  for (size_t i=0; i<data.size(); ++i) {
+   8   	    total += data[i];
+   9   	  }
 ```
+{: data-highlight="1,8" }
+
+### Step out
+Enter `up` AKA "Step Out".  The `sum()` function completes, and the program pauses again.
```
Comment (ohjuny) [inline docs/setup_lldb.md:264]:
```suggestion
Enter `up` (Step Out).  The `sum()` function completes, and the program pauses again.
```
Refer to AKA comment above.

File: docs/setup_lldb.md [hunk 8]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
-   39    
-   40      vector<double> data;
-   41      data.push_back(1);
-Target 0: (stats_tests.exe) stopped.
+    frame #0: 0x0000000100003d7c main.exe`sum(data=size=3) at main.cpp:6:10
+   3   	using namespace std;
+   4   	
+   5   	double sum (const vector<double> &data) {
+-> 6   	  double total = 0;
+   7   	  for (size_t i=0; i<data.size(); ++i) {
```
Comment (ohjuny) [inline docs/setup_lldb.md:257]:
```suggestion
   7   	  for (size_t i = 0; i < data.size(); ++i) {
```
Consistency with nitpick above...

File: docs/setup_lldb.md [hunk 9]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
-Quitting LLDB will kill one or more processes. Do you really want to proceed: [Y/n] y
 ```
+{: data-highlight="1" }
 
-
-# Pro-tips
+## Pro-tips
 Use the up and down arrow keys to cycle through previous commands similar to your command line.
 
 Use TAB completion to automatically complete the name of a command or a variable.
```
Comment (ohjuny) [inline docs/setup_lldb.md:302]:
```suggestion
Use <kbd>TAB</kbd> completion to automatically complete the name of a command or a variable.
```
Styling suggestion.

**Review by ohjuny — COMMENTED** [review]
Generally LGTM, consider modifying commands at the input redirection section in case this is an issue for students.

File: docs/setup_lldb.md [hunk 10]
```diff
@@ -79,93 +50,47 @@ lldb-1200.0.44.2
...
+
+First, compile and run your executable at the command line.
 ```console
-$ pwd
-/Users/awdeorio/src/eecs280/p1-stats
-$ make clean
-rm -rvf *.exe *~ *.out *.dSYM *.stackdump
-$ make stats_tests.exe
-g++ -Wall -Werror -pedantic -g --std=c++11 stats_tests.cpp stats.cpp p1_library.cpp -o stats_tests.exe
+$ make main.exe
```
[the remark this answers is not in the document]
Comment (awdeorio) [reply docs/setup_lldb.md]: Good suggestion.  Take a look at c26e34610691defe4a1a60e46e04be797ba8ef2f

File: docs/setup_lldb.md [hunk 11]
```diff
@@ -79,93 +50,47 @@ lldb-1200.0.44.2
...
-enter a filename
-main_test_data.tsv
-enter a column name
-B
-...
-```
+First, edit your `Makefile` and add the `CXXFLAGS` recommended by the [ASAN Quick Start](setup_asan.html#quick-start).
+
+### Input redirection
+If you're unfamiliar with input redirection, first read the CLI tutorial section on [input redirection](cli.html#input-redirection-).
```
[the remark this answers is not in the document]
Comment (awdeorio) [reply docs/setup_lldb.md:94]: Good catch! Fixed.  30e380a3a8c537b5339b323fc4096cbc1da9cbfd

File: docs/setup_lldb.md [hunk 12]
```diff
@@ -175,156 +100,211 @@ $ lldb main.exe
...
 ```
+{: data-highlight="2" }
+
+## Debug
+In this section, we'll set a breakpoint, which pauses the debugger.  Then, we'll cover some of the options to continue execution.
+
+`n` **Next AKA Step Over**
+Run one line of code, stepping _over_ any function calls by running the whole function in one step.
+
+`s` **Step AKA Step Into**
```
[the remark this answers is not in the document]
Comment (awdeorio) [reply docs/setup_lldb.md]: Fixed!

**Comment by awdeorio** [comment]
Thanks for the detailed review @ohjuny !  I think I've addressed all your comments.

**Comment by ohjuny** [comment]
LGTM 🍏

**Review by ohjuny — APPROVED** [review]
LGTM 🍏

**Pull request merged**